coderd/workspaces_test.go

```diff
@@ -3722,3 +3723,50 @@ func TestWorkspaceTimings(t *testing.T) {
 		})
 	}
 }
+
+func TestGetAuthorizedWorkspacesAndAgents(t *testing.T) {
+	t.Parallel()
+
+	ownerClient, _, api := coderdtest.NewWithAPI(t, &coderdtest.Options{
+		IncludeProvisionerDaemon: true,
+	})
+	owner := coderdtest.CreateFirstUser(t, ownerClient)
+	_, user := coderdtest.CreateAnotherUser(t, ownerClient, owner.OrganizationID)
+	authToken := uuid.NewString()
+	version := coderdtest.CreateTemplateVersion(t, ownerClient, owner.OrganizationID, &echo.Responses{
+		Parse:          echo.ParseComplete,
+		ProvisionPlan:  echo.PlanComplete,
+		ProvisionApply: echo.ProvisionApplyWithAgent(authToken),
+	})
+	coderdtest.AwaitTemplateVersionJobCompleted(t, ownerClient, version.ID)
+	template := coderdtest.CreateTemplate(t, ownerClient, owner.OrganizationID, version.ID)
+	workspace := coderdtest.CreateWorkspace(t, ownerClient, template.ID)
+	coderdtest.AwaitWorkspaceBuildJobCompleted(t, ownerClient, workspace.LatestBuild.ID)
+	_ = agenttest.New(t, ownerClient.URL, authToken)
+	_ = coderdtest.NewWorkspaceAgentWaiter(t, ownerClient, workspace.ID).Wait()
+
+	ctx, cancel := context.WithTimeout(context.Background(), testutil.WaitLong)
+	defer cancel()
+
+	ownerSubject, _, err := httpmw.UserRBACSubject(ctx, api.Database, owner.UserID, rbac.ExpandableScope(rbac.ScopeAll))
+	require.NoError(t, err)
+	preparedOwner, err := api.Authorizer.Prepare(ctx, ownerSubject, policy.ActionRead, rbac.ResourceWorkspace.Type)
+	require.NoError(t, err)
+
+	ownerCtx := dbauthz.As(ctx, ownerSubject)
+	ownerRows, err := api.Database.GetAuthorizedWorkspacesAndAgents(ownerCtx, preparedOwner)
+	require.NoError(t, err)
+	require.Len(t, ownerRows, 1)
+	require.Equal(t, workspace.ID, ownerRows[0].WorkspaceID)
+	require.Len(t, ownerRows[0].AgentIds, 1)
+
+	userSubject, _, err := httpmw.UserRBACSubject(ctx, api.Database, user.ID, rbac.ExpandableScope(rbac.ScopeAll))
+	require.NoError(t, err)
+	preparedUser, err := api.Authorizer.Prepare(ctx, userSubject, policy.ActionRead, rbac.ResourceWorkspace.Type)
+	require.NoError(t, err)
+
+	userCtx := dbauthz.As(ctx, userSubject)
+	userRows, err := api.Database.GetAuthorizedWorkspacesAndAgents(userCtx, preparedUser)
+	require.NoError(t, err)
+	require.Len(t, userRows, 0)
```

Comment:
this test would be more convincing if there were multiple workspaces with multiple builds with multiple agents

Reply:
I'll improve the one in `querier_test` instead